### ISSUE TITLE
[Bug]: How to use text streaming

### ISSUE THREAD

**Issue by stevemao** [microsoft/cognitive-services-speech-sdk-js#850]
### What happened?

On the API docs, I found https://learn.microsoft.com/en-au/azure/ai-services/speech-service/how-to-lower-speech-synthesis-latency?pivots=programming-language-python#text-streaming

But I do not see `speechsdk.SpeechSynthesisRequestInputType.TextStream` in this library. How do achieve text streaming using this library?

### Version

1.36.0 (Latest)

### What browser/platform are you seeing the problem on?

Node

### Relevant log output

```shell
Key steps
Create a text stream request: Use speechsdk.SpeechSynthesisRequestInputType.TextStream to initiate a text stream.

Set global properties: Adjust settings such as output format and voice name directly, as the feature handles partial text inputs and doesn't support SSML. Refer to the following sample code for instructions on how to set them. OpenAI text to speech voices aren't supported by the text streaming feature. See this language table for full language support.

Python

Copy
# set a voice name
speech_config.speech_synthesis_voice_name = "en-US-AvaMultilingualNeural"
Stream your text: For each text chunk generated from a GPT model, use request.input_stream.write(text) to send the text to the stream.

Close the stream: Once the GPT model completes its output, close the stream using request.input_stream.close().
```

**Comment by maisterr** [comment]
Faced the same problem, it would be useful to know if this is planned to be added to the JS SDK soon

**Comment by BrianMouncer** [comment]
@yulin-li can you comment on plans to add TextStreaming to JS SDK?

**Comment by yulin-li** [comment]
Thanks for using the text streaming feature. We need the sync internally for this feature plan. I will reply here once it's clear.

**Comment by akshay-jam** [comment]
@yulin-li any update on when this feature will be available in JS SDK

**Comment by akshay-jam** [comment]
@BrianMouncer @yulin-li  any update on when this feature will be available in JS SDK

**Comment by codylittle** [comment]
Hey @BrianMouncer,

Is this still on the horizon for support? We've currently written our own wrapper, but would prefer to use a SDK with first class support.

Regards,
Cody

**Comment by nanmu42** [comment]
Hi, @codylittle , I am implementing our a wrapper for streaming text-to-speech too. I am trying to talk with `wss://{region}.tts.speech.microsoft.com/cognitiveservices/websocket/v2` using some websocket client directly so I am curious how do you manage to work out what the payload and protocol look like? Any insight would be much appreciated. Thank you.

**Comment by codylittle** [comment]
Hey @nanmu42, we configured an opaque http proxy and monitored the ws traffic. We're in the process of reaffirming our findings with the product team against their internal documentation.

**Comment by nanmu42** [comment]
Thanks @codylittle , very informative. I will try!

I am currently using [Azure Speech Studio](https://speech.microsoft.com/portal) voice gallery to trigger the generation and browse the websocket request in Chrome dev tools. It's based on [the v1 protocol](https://github.com/microsoft/cognitive-services-speech-sdk-js/blob/2c080e58b435e7029eba2b0790b91efd9464efe1/tests/SpeechSynthesisTests.ts#L874), though.

**Comment by nanmu42** [comment]
Hi, @codylittle , sorry to pin you again.

> we configured an opaque http proxy on the python SDK and monitored the ws traffic.

I have also set up a mitmproxy in transparent mode to inspect the websocket traffic. I have installed the mitmproxy CA globally on the python host machine, but it seems that `azure-cognitiveservices-speech 1.42.0` does not trust the CA and declined to actually make the request.

Could you please share your way to inspect the traffic?

Thank you and hope you are well.

**Comment by codylittle** [comment]
Hey @nanmu42, no stress, happy to provide some insights.
We also used mitmproxy, on install it'll create a handful of certificates within your home directory (`~/.mitmproxy`). You'll need to install the preferred format of `mitmproxy-ca-cert` within your trusted roots certificate store.

Once the certs are installed, restart your terminal/IDE

The stream does appear to be a TR-396 (UPS) stream but we replicated it without implementing TR-396.

First class support of this within the SDK would be fantastic, but given the required underlying changes to the transport interfaces, I understand why they haven't yet.

**Comment by MathiasSoender** [comment]
Hi, any updates for this?
Wondering if this functionality will be released anytime soon, as I might go down the road of building a wrapper as well.

Cheers!

**Comment by nanmu42** [comment]
Yes, though I am happy now, it would be really cool if there's a official support in SDK. Azure has really good TTS and ASR. Lacking SDK support is a pity.

**Comment by MathiasSoender** [comment]
@nanmu42 , did you manage to implement the input streaming successfully? Would love to see some example code if possible for you to share! Cheers

**Comment by stevemao** [comment]
Please share the repo. Thanks

**Comment by MathiasSoender** [comment]
Hi, any news on supporting text-input streaming for the Node SDK? Thanks!

**Comment by MathiasSoender** [comment]
For anybody still looking to implement the streaming functionality, I figured a way to do it a "pseudo" fashion.

**Wanted functionality:**

- Input text for the synthesizer arrives as chunks (think some GenAI model piping out it's answer in chunks, which we wish to query instantly to the synthesizer)
- Output audio (the synthesized speech) is sent to the client/user, as chunks. That is, we wish for the time-for-first audio latency to be as small as possible, such that the user can start listening to the audio as soon as possible, and before waiting for the whole text to be synthesized. 

**Idea** (let me know if a full code example is needed):
First operation is to split the generated text chunks into sentences which can be given to the synthesizer. I found the best results by splitting / appending chunks by looking for common delimiters ("! ", ". ", "? "). Example: The sentence, "Hello world! Let's begin." would be split into two sentences; "Hello world!", "Let's begin."

Now the producer loop starts:
Each time, a new sentence is generated, give it to the synthesizer (actually, create a new synthesizer object for each time a sentence needs to be synthesized). This will obviously start multiple processes, which needs to run concurrently.

Now the consumer loop starts:
In the order, which the sentences where given to the synthesizer, one can subscribe to the callback "synthesizer.synthesizing" (that is, a callback for the synthesizer object). This will yield audio chunks as they are generated. Start piping these back to the user. Remember to fully exhaust each of these synthesizer processes, before continuing to the next, otherwise the audio/speech for each sentence might be mixed together.

OR (a more simple solution),

Just wait for the speech to be complete for the first sentence, send this, continue waiting for the second sentence, etc. (Use the "speakTextAsync" method, and send the result when it arrives).


**Results**

A time-for-first audio latency of between 200ms-500ms (commonly around 300ms), using 'en-US-AvaMultilingualNeural'. This is more than enough for my speech service, as the GenAI model I use typically requires about 0.5-1s to generate the first sentence. So, we are looking at max 1.5s in total for a response to be generated.
Obviously, this solution works depending on the nature of the input. With very long sentences being generated by the GenAI model, this wont work.
